---
title: "seqpac - Guide for sRNA analysis"
author: "Daniel NÃ¤tt"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{seqpac - Guide for sRNA analysis}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```
## Introduction

Seqpac contains a complete toolbox for the analysis of short sequenced reads,
where statistical units are counts of unique sequences. The package was
originally developed for small RNA (sRNA) analysis, where unique sequences are
often specific species of sRNA. The seqpac workflow can, however, be applied on
any type of data generated by large scale nucleotide sequencing, where the user
wish to maintain sequence integrity during the whole analysis. This involves, for
example, regular RNA-seq with long RNA.

Thus, seqpac applies sequence-based counting. This can be contrasted against
feature-based counting, where reads are counted over known genomic features,
such a protein coding gene or sRNA. This will result in a count table where a
count is made per feature (e.g. number of read counts mapping to a gene). In
seqpac, such annotations are initially disregarded, as a count table is
generated solely over unique sequences in the raw sequence file. Annotation
against known refereces (such as databases containing miRNA, tRNA, or protein
coding genes) are done after the count table has been generated.

The big advantage of using sequence-based counting, compared to feature-based
counting, is that mismatches in the classification of reads, can be applied
anytime during the analysis. In feature-based counting, allowing a mismatch in
the mapping will result in a loss of information (loss sequence integrity).

For example, say that you have 506 read counts for miR-185 (a miRNA) when
allowing 3 mismatches against the reference. In feature-based analysis this will
result in 1 entery in the count table. Thus you have lost the information about
what sequences that are included in that count. The only thing you know is that
the sequence may differ from the reference sequence by any 3 mismaches. In
sequence-based counting, reads are counted as uniques sequences and each unique
sequence is then mapped againts the target reference databases. Information
about whether a sequence was mapped against a reference is maintained in an
seperate annotation table that can contain both perfect mapping and mapping
allowing mismatches. Thus in any stage of the analysis mismatches and
classification into features can be dynamically controlled, meaning that you can
easily observe the effect of allowing mismatches/classification or not.


## The seqpac workflow

prerequisite
. The pre-requicite  fastq raw sequence
file and

analysis.  sequence-


The core of the seqpac workflow is the
generation and subsequence analysis/visualization of a standardized list object
called PAC. Using an innovative target system, seqpac extracts, analyze and
visualize group differences using the PAC object.

```{r setup}
library(seqpac)
```
