#' Plot stackbars from PAC
#'
#' \code{PAC_stackbar} analyses nucleotide bias.
#'
#' Given a PAC object the function will attempt to extract group information
#' from Pheno, biotype information from Anno, and summaries this over the data
#' in Counts or norm.
#'
#' @family PAC analysis
#'
#' @seealso \url{https://github.com/Danis102} for updates on the current
#'   package.
#'
#' @param PAC PAC-list object.
#' 
#' @param anno_target List with: 
#'                          1st object being character vector of target column(s) in Anno, 
#'                          2nd object being a character vector of the target
#'                          biotype(s) in the target column (1st object).
#'                          Important, the 2nd object is order sensitive,
#'                          meaning that categories will appear in the same
#'                          order in the stacked bargraph. (default=NULL)
#'
#'
#' @param pheno_target List with: 
#'                          1st object being character vector of target column(s) in Pheno, 
#'                          2nd object being a character vector of the target
#'                          group(s) in the target column (1st object).
#'                          Important, the 2nd object is order sensitive,
#'                          meaning that categories will appear in the same
#'                          order in the stacked bargraph. (default=NULL)
#'                          
#' @return A stacked bar plot generated by ggplot2 
#'
#' @examples
#' 
#' path="/data/Data_analysis/Projects/Drosophila/Other/IOR/Joint_analysis/R_analysis/"
#' load(file=paste0(path, "PAC_all.Rdata"))
#' 
#' PAC_filt <- PAC_filter(PAC_all, size=c(16,45), threshold=10, coverage=5, type="counts", stat=FALSE, pheno_target=NULL, anno_target=NULL)
#' PAC_filt <- PAC_rpm(PAC_filt)
#' PAC_filt <- PAC_filter(PAC_filt, size=c(16,45), threshold=10, coverage=5, type="rpm", stat=FALSE, pheno_target=NULL, anno_target=NULL)
#' 
#' hierachy <- list( Mt_rRNA= "12S|16S|Mt_rRNA",
#'                  rRNA="5S|5.8S|18S|28S|S45|Ensembl_rRNA|rRNA_Other",
#'                  Mt_tRNA= "tRNA_mt-tRNA",
#'                  tRNA="Ensembl_tRNA|tRNA_nuc-tRNA",
#'                  miRNA="^miRNA|Ensembl_miRNA|Ensembl_pre_miRNA",
#'                  piRNA="piRNA")
#' 
#' PAC_filt <- simplify_reanno(PAC_filt, hierachy=hierachy, mismatches=0, bio_name="Biotypes_1", PAC_merge = TRUE)
#' 
#' PAC_filt$Pheno$Groups <- paste(do.call("rbind", strsplit(as.character(PAC_filt$Pheno$SampleProject), "_" ))[,1], PAC_filt$Pheno$Method, PAC_filt$Pheno$Method_tag, PAC_filt$Pheno$Tag, sep="_")
#' 
#' 
#' biotypes_ord <- c("rRNA","Mt_rRNA", "tRNA", "Mt_tRNA", "miRNA", "piRNA","other", "no_anno")
#' method_ord <- c("Sep_Proto_Reg__", "Sep_TGIRT_Reg__", "Sep_Proto_IOR__", "Jan_Proto_IOR__", 
#'                 "Sep_Proto_Long_Tag", "Sep_Proto_Short_Tag", "Jan_Proto_Long_Tag", "Jan_Proto_Short_Tag",
#'                 "Sep_Proto_Long_No_Tag", "Sep_Proto_Short_No_Tag", "Jan_Proto_Long_No_Tag", "Jan_Proto_Short_No_Tag",
#'                 "Sep_TGIRT_IOR__", "Jan_TGIRT_IOR__", "Jan_R2_R2R_TGIRT_R2_R2R__")
#' 
#' plot  <- PAC_stackbar(PAC_filt, anno_target = list("Biotypes_1", biotypes_ord), pheno_target = list("Groups", method_ord), width=1, no_anno=TRUE)  
#' 
#' 
#' table(PAC_IOR1$Anno$Biotypes_simpl)
#' 
#' 
#' @export

PAC_stackbar <- function(PAC, anno_target=list("Biotypes", NULL), pheno_target=NULL, colvec=NULL, width=1.0, no_anno=TRUE){
                                                require(ggthemes)
                                                require(ggplot2)
                                                
                                                ### Generate input data
                                                stopifnot(identical(rownames(PAC$Pheno), colnames(PAC$Counts)))
                                                stopifnot(identical(rownames(PAC$Anno), rownames(PAC$Counts)))
                                                data <- PAC$Counts
                                                anno <- PAC$Anno
                                                if(no_anno==FALSE){
                                                          data <- data[!as.character(anno[, anno_target[[1]]]) == "no_anno",]
                                                          anno <- anno[!as.character(anno[, anno_target[[1]]]) == "no_anno",]
                                                          }
                                                tot_cnts <- colSums(data)
                                                data_shrt <- aggregate(data, list(anno[, anno_target[[1]]]), "sum")
                                                data_shrt_perc <- data_shrt
                                                data_shrt_perc[,-1] <- "NA"
                                                for (i in 1:length(tot_cnts)){ 
                                                                    data_shrt_perc[,1+i]   <- data_shrt[,1+i]/tot_cnts[i]
                                                }
                                                
                                                data_long_perc <- reshape2::melt(data_shrt_perc, id.vars="Group.1")
                                                colnames(data_long_perc) <- c("Biotype", "Sample", "Percent")
                                                if(!is.null(anno_target[[2]])){
                                                          data_long_perc$Biotype <- factor(as.character(data_long_perc$Biotype), levels=rev(anno_target[[2]]))}
                                                data_long_perc$Percent <- data_long_perc$Percent*100 
                                                
                                                ### Fix sample order
                                                if(is.null(pheno_target)){
                                                        data_long_perc$Sample <- factor(as.character(data_long_perc$Sample), levels=as.character(unique(data_long_perc$Sample)))
                                                }else{  
                                                        stopifnot(any(!rownames(PAC$Pheno) %in% as.character(data_long_perc$Sample))==FALSE)
                                                        sampl_ord <- do.call("c", split(rownames(PAC$Pheno), factor(PAC$Pheno[,pheno_target[[1]]], levels=pheno_target[[2]])))
                                                        data_long_perc$Sample <- factor(as.character(data_long_perc$Sample), levels=as.character(sampl_ord))
                                                }
                                                        
                                                
                                                
                                                ### Plot data
                                                if(is.null(colvec)){
                                                          bio <- as.character(unique(data_long_perc$Biotype))
                                                          n_extra  <- sum(bio %in% c("no_anno", "other"))
                                                          colfunc <- colorRampPalette(c("#094A6B", "#FFFFCC", "#9D0014"))
                                                          if(n_extra==1){colvec <- c(colfunc(length(bio)-1), "#6E6E6E")}
                                                          if(n_extra==2){colvec <- c(colfunc(length(bio)-2), "#6E6E6E", "#BCBCBD")}
                                                          if(n_extra==0){colvec <- colfunc(length(bio))}
                                                          }
                                                p1<- ggplot(data_long_perc, aes(x=Sample, y=Percent, fill=Biotype)) +
                                                                              geom_bar(stat="identity", width=width, size=1.0) + 
                                                								              geom_hline(yintercept=0, col="#707177", cex=0.6) +
                                                                              geom_rangeframe(aes(y=c(0, rep(1, length(Percent)-1))))+
                                                  												    scale_fill_manual(values=rev(colvec))+
                                                                              ylab("Percent of total reads")+
                                                                              theme_tufte()+
                                                                              theme(
                                                                                  #text = (element_text(family="helvetica")),
                                                                  					      plot.caption =  element_text(size=12, face= "bold"),
                                                                  					      axis.title.y = element_text(size=16, face= "bold"), 
                                                                  					      axis.title.x = element_blank(), 
                                                                  					      axis.text = element_text(size=14),
                                                                  					      axis.text.x = element_text(angle=45, hjust=1),
                                                                  					      panel.background = element_blank()) 
                                                return(p1)
                                                }