#' Genereate snRNA Progress Report from Sports
#'
#' \code{progress.report} genereates summarized snRNA progress report
#' from Sports intermediate files.
#'
#' Given the path to Sports output directory, this function will import all
#' progress report txt.files of a given Sports project and summarize them into a
#' single dataframe. The progress report dataframe can then easily be added to
#' the Pheno object in the PAC file for further analysis.
#'
#' @section Important: Progress reports will only be generated if using the -k
#'   option in Sports.
#'
#' @family PAC generation
#'
#' @seealso \url{https://github.com/junchaoshi/sports1.0} for download and
#'   documentation about running Sports. \url{https://github.com/Danis102} for
#'   updates on the current package.
#'
#' @param path The Sport output path. The function will search this path for the
#'   the progress_report folder that is automatically generated by Sports
#'   running with the -k option (save all intermediate files).
#'
#' @return data.frame
#'
#' @examples
#' report <- progress_report(<your_path_to_sports_output_directory>)
#' report <- progress_report(path=path)
#'
#' @export

##--------------------------------------------------------------------------------##
# Only works if running Sports with -k (save all intermediate files).
progress_report <- function(path){
                        ### Get paths and sample names for output
                        meta_files <- list.files(paste0(path, "/processing_report"), pattern ="*.txt", full.names=TRUE)
                        meta_files_nams <- list.files(paste0(path, "/processing_report"), full.names=FALSE)
                        meta_files_nams <- gsub("\\<1_|_merge.txt", "", meta_files_nams)
                        meta_files_nams <- gsub("-", "_", meta_files_nams)
                        ### Generate colnames in output from first file
                        colnam <- readLines(meta_files[[1]]) # Look in one to find out column numbers
                        colnam <- gsub("match to |_genome", "", colnam[grepl("match to", colnam)])
                        colnam <- colnam[!grepl("database", colnam)]
                        ### Generate empty output file
                        df <- data.frame(matrix(NA, nrow=length(meta_files), ncol=length(colnam)))
                        rownames(df) <- meta_files_nams
                        colnames(df) <-  colnam
                        for (i in 1:length(meta_files)){
                                lines <- readLines(meta_files[[i]])
                                nam <-  meta_files[[i]]
                                fact <- cumsum(!nzchar(lines)) # Cumulativly/progressively sums TRUE argumentsfor for empty lines
                                split_lst <- split(lines, fact)
                                split_lst <- lapply(split_lst, function(x){ y <- x[nzchar(x)]; return(y)}) # Removes empty lines with no characters
                                split_lst_red <- split_lst[do.call("c", lapply(split_lst, function(x){!length(x)==1}))] # Removes headline strings with no data
                                split_lst_red <- split_lst_red[grepl("match to", split_lst_red)] # Removes last line without metadata
                                test <- gsub("match to |_genome", "", do.call("c", lapply(split_lst_red, function(x){x[grepl("match to", x)]})))
                                if(!identical(as.character(colnam), as.character(test))){stop("Error: colnames in output dataframe imported Sports meta file does not match up.\n
                                                                                              Have you used different mapping databases for different samples?")}
                                data <- do.call("c", lapply(split_lst_red, function(x){x[grepl("# reads with at least one reported alignment: ", x)]}))
                                df[i,] <- gsub("# reads with at least one reported alignment: ", "", data)
                        }
                        return(df)
}
